File: sort/2750QS.py
n=int(input())
ls= [int(input()) for _ in range(n)]

def QS(a, b):
    if a>=b: return
    start =a
    end= b
    p= ls[(start+end)//2]
    while start<=end:
        while ls[start]<p: start+=1
        while ls[end]>p:  end-=1
        if start<=end:
            ls[start], ls[end]= ls[end], ls[start]
            start+=1
            end-=1

    QS(a, end)
    QS(start, b)

QS(0,n-1)

[print(el) for el in ls]

File: sort/2750QS.cpp
#include <stdio.h>
using namespace std;
#define FOR(i,n) for(int i=0;i<n;++i)
void QS(int start, int end);
int ls[1000+1];

int main()
{
    int n;
    scanf("%d",&n);
    FOR(i,n){
        int x;
        scanf("%d", &x);
        ls[i]=x;
    }
    QS(0,n-1);
    FOR(i,n) printf("%d\n",ls[i]);
}

void swap(int* a, int* b)
{
    int tmp= *a;
    *a= *b;
    *b=tmp;
}

/*
inline void swap(int &i, int &j)
{
    int temp= i;
    i= j;
    j= temp;
}

inline void swap(double &i, double &j)
{
    double temp= i;
    i= j;
    j= temp;
}
*/
void QS(int a, int b)
{
    if(a>=b) return;
    int start =a;
    int end= b;
    int p= ls[(start+end)>>1];
    while(start<=end)
    {
        while(ls[start]<p) ++start;
        while(ls[end]>p) --end;
        if(start<=end){
            if(ls[start]>ls[end])
                swap(&ls[start],&ls[end]);
            ++start;
            --end;
        }
    }
    QS(a, end);
    QS(start, b);
}